src/parse.rs

```diff
@@ -99,20 +104,57 @@ impl<'a> core::iter::FusedIterator for HexToBytesIter<'a> {}
 
 #[cfg(any(test, feature = "std", feature = "alloc"))]
 impl FromHex for Vec<u8> {
-    fn from_byte_iter<I>(iter: I) -> Result<Self, Error>
+    type Error = HexToBytesError;
+
+    fn from_byte_iter<I>(iter: I) -> Result<Self, HexToBytesError>
     where
-        I: Iterator<Item = Result<u8, Error>> + ExactSizeIterator + DoubleEndedIterator,
+        I: Iterator<Item = Result<u8, HexToBytesError>> + ExactSizeIterator + DoubleEndedIterator,
     {
         iter.collect()
     }
 }
 
+/// Hex decoding error.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum HexToBytesError {
+    /// Non-hexadecimal character.
+    InvalidChar(u8),
```

Comment:
I tried putting `#[non_exhaustive]` on enum error variants so we can do future changes like you mention but it seems to make the error type basically unusable because then one cannot match on the variants.

Reply:
You can definitely match on `#[non_exhaustive]` errors. You just need to stick a `_` clause on to deal with potential future errors.

Having said this, `#[non_exhaustive]` would be insufficient to *replace* one variant with another, or to change its internal fields.

@Kixunil when we return these errors from iterators we may not know the "position", and accumulating multiple erroneous chars would require allocation. I'm not sure it's worth it for an error variant like this, where we give the user a character they can easily search their string for (and where the error will almost always indicate "this string is not hex at all, exact error is irrelevant").

Reply:
> You can definitely match on #[non_exhaustive] errors. You just need to stick a _ clause on to deal with potential future errors.
>
> Having said this, #[non_exhaustive] would be insufficient to replace one variant with another, or to change its internal fields.

Oh I meant putting `#[non_exhaustive]` on each variant. This idea came from that error blog post you linked to a while back.

Reply:
Ah, yeah, I see. Agreed, I'd rather not do this. We'd have to implement a ton of accessor methods and stuff. Let's just get it right the first time :)

Reply:
We may not know the position but we can use `Option<usize>` for it. (Although I believe we always can know it.)

Yes, reporting multiple errors should be opt-in because of required allocation. I've started working on something that looks pretty neat to me. What I have in mind is rustc-like beautiful errors.